R/compute_weighted_intensity_levels.R

```diff
@@ -0,0 +1,152 @@
+#'      - 'lnorm': Uses the Log-normal distribution for fitting.
+#'      - 'exp': Uses the Exponential distribution for fitting.
+#'
+#' @export
+#'
+#' @examples
+#' # Create three seasons with random observations
+#' obs = 10
+#' season = c("2018/2019", "2019/2020", "2020/2021")
+#' season_num_rev <- rev(seq(from = 1, to = length(season)))
+#' observations = rep(stats::rnorm(10,obs), length(season))
+#'
+#' # Add into a tibble with weight decreasing going one season step back
+#' data_input <- tibble::tibble(
+#'   observation = observations,
+#'   weight = 0.8^rep(season_num_rev, each = obs)
+#' )
+#'
+#' # Use the model
+#' compute_weighted_intensity_levels(weighted_observations = data_input,
+#'                                   conf_levels = c(0.50, 0.90, 0.95),
+#'                                   family = "weibull")
+
+compute_weighted_intensity_levels <- function(
+  weighted_observations,
+  conf_levels = c(0.50, 0.90, 0.95),
+  family = c("weibull",
+             "lnorm",
+             "exp"),
+  optim_method = c("Nelder-Mead",
+                   "BFGS",
+                   "CG",
+                   "L-BFGS-B",
+                   "SANN",
+                   "Brent"),
+  lower_optim = -Inf,
+  upper_optim = Inf
+) {
+  # Check input arguments
+  coll <- checkmate::makeAssertCollection()
+  checkmate::assert_tibble(weighted_observations, add = coll)
+  checkmate::assert_numeric(conf_levels, lower = 0, upper = 1, len = 3,
+                            unique = TRUE, sorted = TRUE, add = coll)
+  checkmate::assert_names(colnames(weighted_observations),
+                          identical.to = c("observation", "weight"), add = coll)
+  checkmate::assert_numeric(lower_optim, add = coll)
+  checkmate::assert_numeric(upper_optim, add = coll)
+  checkmate::reportAssertions(coll)
+  # Match the arguments.
+  family <- rlang::arg_match(family)
+  optim_method <- rlang::arg_match(optim_method)
+
+  # Initialising parameters based on family
+  init_par_fun <- function(family, observations) {
+    init_params <- switch(family,
+      weibull = log(c(1.5, mean(observations))),
+      lnorm = c(mean(log(observations)), stats::sd(log(observations))),
+      exp = log(1.5)
+    )
+    return(init_params)
+  }
+
+  # The weighted negative loglikelihood function
+  nll <- function(par, weighted_observations, family = family) {
+    switch(family,
+      weibull = -sum(stats::dweibull(weighted_observations$observation, shape = exp(par[1]), scale = exp(par[2]),
+                                     log = TRUE) * weighted_observations$weight),
+      lnorm = -sum(stats::dlnorm(weighted_observations$observation, meanlog =  par[1], sdlog = par[2],
+                                 log = TRUE) * weighted_observations$weight),
+      exp =  -sum(stats::dexp(weighted_observations$observation, rate = exp(par[1]),
+                              log = TRUE) * weighted_observations$weight)
+    )
+  }
```

Comment:
I would clean this up a bit to make it easier to read. 
My suggestion:

```suggestion
  # The weighted negative loglikelihood function
  nll <- function(par, weighted_observations, family = family) {
    log_probability <- switch(family,
      weibull = stats::dweibull(weighted_observations$observation, shape = exp(par[1]), scale = exp(par[2]), log = TRUE),
      lnorm = stats::dlnorm(weighted_observations$observation, meanlog =  par[1], sdlog = par[2], log = TRUE),
      exp = stats::dexp(weighted_observations$observation, rate = exp(par[1]), log = TRUE) 
    )
    
    return(-sum(log_probability * weighted_observations$weight))
  }
```